lib/minheap.js

```diff
@@ -38,26 +46,50 @@ class MinHeap {
   }
 
   // This method returns true if the heap is empty
-  // Time complexity: ?
-  // Space complexity: ?
+  // Time complexity: O(1)
+  // Space complexity: O(1)
   isEmpty() {
-    throw new Error("Method not implemented yet...");
+    return this.store.length === 0;
   }
 
   // This helper method takes an index and
   //  moves it up the heap, if it is less than it's parent node.
   //  It could be **very** helpful for the add method.
-  // Time complexity: ?
-  // Space complexity: ?
+  // Time complexity: O(log n)
+  // Space complexity: O(1)
   heapUp(index) {
-    throw new Error("Method not implemented yet...");
+    let index1 = index;
+    let parentIndex = Math.floor((index - 1)/2);
+
+    while (parentIndex >= 0 && this.store[parentIndex].key > this.store[index1].key) {
+      this.swap(index1, parentIndex);
+      index1 = parentIndex;
+      parentIndex = Math.floor((index1 - 1)/2);
+    }
   }
 
   // This helper method takes an index and 
   //  moves it up the heap if it's smaller
   //  than it's parent node.
   heapDown(index) {
-    throw new Error("Method not implemented yet...");
+    let leftChildIndex = (2 * index) + 1;
+    let rightChildIndex = (2 * index) + 2;
+    let nextRootIndex;
+
+    if (leftChildIndex < this.store.length && rightChildIndex < this.store.length) {
+      if (this.store[leftChildIndex].key > this.store[rightChildIndex].key) {
+        nextRootIndex = rightChildIndex;
+      } else {
+        nextRootIndex = leftChildIndex;
+      }
+    } else {
+      nextRootIndex = leftChildIndex;
+    }
```

Comment:
If we initialize next to be the more common outcome of the conditions, we can reduce the complexity of the conditional clauses. Further, we can observe that the right child being in bounds implies the left must be as well and omit that from the condition.

```js
    let nextRootIndex = leftChildIndex;

    if (rightChildIndex < this.store.length) {
      if (this.store[leftChildIndex].key > this.store[rightChildIndex].key) {
        nextRootIndex = rightChildIndex;
      }
    }
```